gtk4: new recipe

tschoonj opened the pull request:
@ryanvolz @pkgw are you both ok being co-maintainers of this recipe?

<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team][1] using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel][2] if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/gtk4```) and found it was in an excellent condition.

pkgw commented:
Yes, I'm willing to help maintain.

pkgw commented on recipes/gtk4/meta.yaml:
```diff
@@ -0,0 +1,137 @@
+build:
+  number: 0
+  # workaround for bug in LIEF (https://github.com/lief-project/LIEF/issues/239)
+  # that resulted in missing symbols
```
I believe this comment about LIEF should go with the following line of YAML:

```
rpaths_patcher: patchelf  # [linux]
```

If this item is intentionally being removed for this recipe, the comment should be removed.

tschoonj replied:
I removed the comment. Thanks for bringing it to my attention!

ryanvolz commented:
> @ryanvolz @pkgw are you both ok being co-maintainers of this recipe?

Yes, I'm happy to help!

tschoonj commented on recipes/gtk4/conda_build_config.yaml:
```diff
@@ -0,0 +1,6 @@
+pango:
+  - 1.48
```
ah great! where is that PR?

tschoonj replied:
`conda-forge-pinning` has already been updated with this change, so can I drop the pin now already?

ryanvolz replied:
It's just a migrator for now, and those don't get applied on staged-recipes. I don't know if you can manually apply the migrator instead of using `conda_build_config.yaml` (I don't think so?). Still fine to have this I think, but we can remove it immediately once the feedstock is created.

ryanvolz commented:
https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/1371 Praise the bot!

tschoonj commented:
There is already a package called `libsass` on conda-forge which comes with an `sassc` script, but is unfortunately incompatible.

It looks like [libsass](https://github.com/sass/libsass) and [sassc](https://github.com/sass/sassc) projects are both deprecated projects, so I am not sure it would be wise to create new feedstocks for them.

It might make cross-compilation a bit more tricky, but nothing that can't be done I think.

ryanvolz commented on recipes/gtk4/meta.yaml:
```diff
@@ -0,0 +1,136 @@
+    - fribidi
+    - freetype
+    - fontconfig
+    - libcups  # [linux]
```
I think we need a pin here since `libcups` doesn't have a run_export (apparently).
```suggestion
    - {{ pin_compatible('libcups') }}  # [linux]
```

ryanvolz commented:
> There is already a package called `libsass` on conda-forge which comes with an `sassc` script, but is unfortunately incompatible.
> 
> It looks like [libsass](https://github.com/sass/libsass) and [sassc](https://github.com/sass/sassc) projects are both deprecated projects, so I am not sure it would be wise to create new feedstocks for them.
> 
> It might make cross-compilation a bit more tricky, but nothing that can't be done I think.

If other reviewers are fine with it, then let's just see how it goes as is. The `gtk` developers have already lightly forked these for their own use and that seems likely to continue/grow given that the parent projects are deprecated. I'm not too inclined to step into that mess if we don't have to.

tschoonj commented:
I added `freetype` and `fontconfig` explicitly to get rid of some warnings, but they are not required for the post-install step (I changed the comment to make that clear).

tschoonj commented:
Ready for a merge?

ryanvolz commented:
Good with me. @conda-forge/staged-recipes This is ready for review.

pkgw reviewed (approved):
This is looking good to me. Very nice work!

chrisburr commented on recipes/gtk4/build.sh:
```diff
@@ -0,0 +1,45 @@
+    -Dwayland-backend=false
+)
+
+if test $(uname) == 'Darwin' ; then
```
It doesn't matter for now as we don't cross-compile between OSs but it's slightly better to use something like `if [[ "${target_platform}" == "osx-"* ]]; then` (and the same for Linux slightly lower down)

chrisburr commented on recipes/gtk4/meta.yaml:
```diff
@@ -0,0 +1,138 @@
+  ignore_run_exports_from:
+    - {{ compiler('cxx') }}
+  missing_dso_whitelist:
+    - '*/libX11.so*'  # [linux]
```
Is there a reason to not include the conda-forge x11 packages as dependencies?

tschoonj replied:
I never tried doing that. Are their counterparts for all these cdt's?

ryanvolz replied:
For reference, the origin of this is https://github.com/conda-forge/staged-recipes/pull/10979/commits/c9a7509d84b39c5373946c8de78639215ed16ea1 from when `gtk3` went through staged-recipes.

The conda-forge x11 vs CDT discussion has come up a few times, notably https://github.com/conda-forge/gtk3-feedstock/issues/7 and https://github.com/conda-forge/cfep/pull/7. We ultimately went with CDTs for `gtk3` for now because that's what `qt` was doing. It has certainly not been headache-free, but I can't speak to potential drawbacks of depending on the conda-forge x11 packages.

chrisburr commented on recipes/gtk4/yum_requirements.txt:
```diff
@@ -0,0 +1,9 @@
+libX11
```
Should we be using the conda-forge packages and/or possibly CDTs for these?

ryanvolz replied:
Can these be replaced by CDTs (now)? I know they're here for the test phase to work and the [docs](https://conda-forge.org/docs/maintainer/knowledge_base.html#core-dependency-tree-packages-cdts) say to put things in `yum_requirements.txt` for that purpose, but it would definitely be cleaner otherwise.

tschoonj replied:
I replaced the CDTs with conda-forge equivalents and that seems to work. I also got rid of the `yum_requirements.txt` file.

ryanvolz replied:
That swath of packages adds 13 MB to the download (tested locally), 11 MB of which is `mesalib`. I'm not entirely sure that the entire `mesalib` is required or what we want. But if we could trim that down it would be good.

ryanvolz replied:
Also if we decide on depending on the xorg- packages here, it probably makes sense to do the same in `gtk3`.

chrisburr replied:
I'm happy either way, I'm happy to merge this once you have a consensus between yourselves.

According to the linking checks `mesalib` isn't being used so it can probably be dropped from run:

```log
WARNING (gtk4): dso library package conda-forge::xorg-libice-1.0.10-h7f98852_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (gtk4): run-exports library package conda-forge::freetype-2.10.4-h0708190_1 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (gtk4): dso library package conda-forge::xorg-libsm-1.2.3-hd9c2040_1000 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (gtk4): dso library package conda-forge::mesalib-18.3.1-h590aaf7_0 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (gtk4): run-exports library package conda-forge::xorg-libxinerama-1.1.4-h9c3ff4c_1001 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (gtk4): dso library package conda-forge::xorg-libxtst-1.2.3-h516909a_1002 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
WARNING (gtk4): dso library package conda-forge::xorg-libxrender-0.9.10-h516909a_1002 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```

ryanvolz replied:
Half the xorg stuff is getting installed with the `gtk3`-dependent programs already and no one has complained about performance issues, so I'm good with just moving to them over the CDTs here (we can think about doing the same with `gtk3` separately). That way the xorg dependencies will be pulled in by any packages that use `gtk4` and there will be no headaches about remembering to add the CDTs to those recipes.
